src/store/actions/submissionActions.js

```diff
@@ -0,0 +1,136 @@
+
+export const uploadWarmup = fileID => ({
+  type: ActionType.UPLOAD_WARMUP,
+  fileID: fileID
+});
+
+export const uploadCanceled = fileID => ({
+  type: ActionType.FILE_UPLOAD_CANCELLED,
+  fileID: fileID
+});
+
+export const uploadStart = (fileID, fName) => ({
+  type: ActionType.UPLOAD_STARTED,
+  fileID: fileID,
+  fileName: fName
+});
+export const uploadProgress = (fileID, progress) => ({
+  type: ActionType.UPLOAD_PROGRESS,
+  fileID: fileID,
+  progress: progress
+});
+
+export const uploadComplete = fileID => ({
+  type: ActionType.UPLOAD_COMPLETE,
+  fileID: fileID
+});
+
+export const uploadFailed = fileID => ({
+  type: ActionType.UPLOAD_FAILED,
+  fileID: fileID
+});
+
+export const clearUpload = () => ({
+  type: ActionType.CLEAR_UPLOAD
+});
+
+export const addPendingFile = fileID => {
+  return async dispatch => {
+    dispatch(uploadWarmup(fileID));
+  };
+};
+
+export const cancelFileUpload = fileID => {
+  return async dispatch => {
+    dispatch(uploadCanceled(fileID));
+  };
+};
+
+export const uploadprocess = (file, fileID, history) => {
+  return async (dispatch, getState) => {
+    const fileTracker = getState().upload.uploadFiles;
+    dispatch(uploadStart(fileID, file.name));
+    let config = {
+      headers: {
+        Authorization: "Bearer " + getState().auth.jwt
+      }
+    };
+    const fileUpload = axios.post(
+      UrlJoin(process.env.REACT_APP_API_URL, `upload`),
+      { fileName: `${getState().hack.hackDetails._id}/${file.name}` },
+      config
+    );
+    fileUpload
+      .then(res => {
+        const url = res.data.fileUploadURL;
+        const config = {
+          headers: { "Content-Type": "multipart/form-data" },
+          onUploadProgress: function(progressEvent) {
+            let percentCompleted = Math.round(
+              (progressEvent.loaded * 100) / progressEvent.total
+            );
+            dispatch(uploadProgress(fileID, percentCompleted));
+          }
+        };
+        axios
+          .put(url, file, config)
+          .then(() => {
+            dispatch(uploadComplete(fileID));
+          })
+          .then(() => {
+            if (fileTracker.indexOf(fileID) === fileTracker.length - 1) {
+              history.push(`/hack/${getState().hack.hackDetails._id}`);
+            }
+          })
+          .catch(error => {
+            dispatch(uploadFailed(fileID));
+            console.log(error);
+          });
+      })
+      .catch(error => {
+        dispatch(uploadFailed(fileID));
+        console.log(error);
+      });
+  };
+};
+
+export const submitHackIdea = (submitData, history) => {
+  return async (dispatch, getState) => {
+    let config = {
+      headers: {
+        Authorization: "Bearer " + getState().auth.jwt
+      }
+    };
+    let files = [];
+    for (let file in submitData.files) {
+      files.push({
+        name: submitData.files[file].name,
+        size: submitData.files[file].size,
+        type: submitData.files[file].type
```

Comment:
why do we need to tell the backend about these two? why do we need to store them in the db at all? it should tell us after the upload has finished by sourcing the data from s3.

Reply:
The file size is reported by s3 however the type is not. The reason behind telling the backend about this is when the user wants to view the submission details the file size and type can be displayed to them.

Reply:
im not saying we should tell the user those details, I'm saying we shouldn't trust the client to give us the right ones. different browsers may give different types for the same type of file (eg, pdfs have been identified as `application/pdf` and `application/x-pdf` over the years, older browsers may still use the x-pdf one).

https://www.npmjs.com/package/file-type or https://www.npmjs.com/package/mime-types might solve the problem of identify the type of file based on the file itself and not the clients lies.

Reply:
Ohh right okay, I was unaware that browsers could report the wrong file type I know different OS's can report different file sizes. I'll make a note of the package however for now I will remove the file type for now as we are not using it in the immediate future.